core/src/com/unciv/logic/automation/BarbarianAutomation.kt

```diff
@@ -0,0 +1,224 @@
+        val enemies = nearEnemyTiles.mapNotNull { it.tileToAttack.militaryUnit }
+        var furthestTile: Pair<TileInfo, Float> = possibleTiles.random() to 0f
+        for (enemy in enemies) {
+            for (tile in possibleTiles) {
+                val distance = enemy.movement.getMovementCostBetweenAdjacentTiles(enemy.currentTile, tile, enemy.civInfo)
+                if (distance > furthestTile.second) {
+                    furthestTile = tile to distance
+                }
+            }
+        }
+        return furthestTile.first
+    }
+
+    private fun healUnit(unit: MapUnit, unitDistanceToTiles: PathsToTilesWithinTurn) {
+        val currentUnitTile = unit.getTile()
+        val bestTilesForHealing = unitDistanceToTiles.keys
+                .filter { unit.movement.canMoveTo(it) }
+                .groupBy { unit.rankTileForHealing(it) }
+                .maxBy { it.key }
+
+        // within the tiles with best healing rate, we'll prefer one which has the highest defensive bonuses
+        val bestTileForHealing = bestTilesForHealing?.value?.maxBy { it.getDefensiveBonus() }
+        if (bestTileForHealing != null
+            && currentUnitTile != bestTileForHealing
+            && unit.rankTileForHealing(bestTileForHealing) > unit.rankTileForHealing(currentUnitTile)) {
+            unit.movement.moveToTile(bestTileForHealing)
+        }
+
+        unit.fortifyIfCan()
+    }
+
+    private fun tryUpgradeUnit(unit: MapUnit, unitActions: List<UnitAction>): Boolean {
+        if (unit.baseUnit().upgradesTo != null) {
+            val upgradedUnit = unit.civInfo.gameInfo.ruleSet.units[unit.baseUnit().upgradesTo!!]!!
+            if (upgradedUnit.isBuildable(unit.civInfo)) {
+                val upgradeAction = unitActions.firstOrNull { it.name.startsWith("Upgrade to") }
+                if (upgradeAction != null && upgradeAction.canAct) {
+                    upgradeAction.action?.invoke()
+                    return true
+                }
+            }
+        }
+        return false
+    }
+
+    private fun tryPillageImprovement(
+            unit: MapUnit,
+            unitDistanceToTiles: PathsToTilesWithinTurn,
+            unitActions: List<UnitAction>
+    ): Boolean {
+        val tilesThatCanWalkToAndThenPillage = unitDistanceToTiles
+                .filter { it.value.totalDistance < unit.currentMovement }.keys
+                .filter { unit.movement.canMoveTo(it) && UnitActions().canPillage(unit, it) }
+
+        if (tilesThatCanWalkToAndThenPillage.isEmpty()) return false
+        val tileToPillage = tilesThatCanWalkToAndThenPillage.maxBy { it.getDefensiveBonus() }!!
+        if (unit.getTile() != tileToPillage) {
+            unit.movement.moveToTile(tileToPillage)
+        }
+
+        unitActions.first { it.name == "Pillage" }.action?.invoke()
+        return true
+    }
+
+    private fun wander(unit: MapUnit, unitDistanceToTiles: PathsToTilesWithinTurn) {
```

Comment:
Aren't there non-barbarian units who random walk? Same question

Reply:
yeah, the same, will use from automation